library/cmac.c

```diff
@@ -867,11 +867,12 @@ static int cmac_test_wth_cipher( int verbose,
         {
             /* When CMAC is implemented by an alternative implementation, or
              * the underlying primitive itself is implemented alternatively,
-             * AES-192 may be unavailable. This should not cause the selftest
-             * function to fail. */
+             * AES-192 and/or 3DES may be unavailable. This should not cause
+             * the selftest function to fail. */
             if( ( ret == MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED ||
                   ret == MBEDTLS_ERR_CIPHER_FEATURE_UNAVAILABLE ) &&
-                  cipher_type == MBEDTLS_CIPHER_AES_192_ECB ) {
+                ( cipher_type == MBEDTLS_CIPHER_AES_192_ECB ||
+                  cipher_type == MBEDTLS_CIPHER_DES_EDE3_ECB ) ) {
```

Comment:
This change makes a difference in two cases:

* If `MBEDTLS_DES_C` and `MBEDTLS_DES_ALT` are enabled, but the DES ALT implementation does not implement 3DES encryption. This is not officially supported in Mbed TLS and does not really make sense (unless you only want to use DES to decrypt legacy single-DES-encrypted data?).
* If `MBEDTLS_DES_C` and `MBEDTLS_CMAC_ALT` enabled, but the CMAC ALT implementation does not support DES. I think that in this case, it should provide a software fallback, or else require `MBEDTLS_DES_C` to be turned off. However, given the low importance of DES today, I'm personally ok with allowing CMAC ALT not to support DES.

I do think this should be documented in `cmac.h`, however. If `MBEDTLS_DES_C` is turned on but not `MBEDTLS_DES_ALT`, which is often not the case for applications running on the kind of platforms that have a hardware accelerator these days, it is somewhat surprising that CMAC wouldn't work with it. Especially if DES has an alt implementation.

Also, note that it's a defect that the CMAC module doesn't support CAMELLIA and ARIA. We may get away with this patch until the migration from ALT implementations to PSA drivers is complete, or we may have to start requiring a fallback mechanism in CMAC ALT implementations.

Reply:
Sorry for letting this sit for so long, but I think your comment might've been confused, @gilles-peskine-arm? This PR does not touch anything functional in the CMAC module itself, it only allows the self-test function to accept that a CMAC_ALT implementation may return `FEATURE_UNSUPPORTED` or `FEATURE_UNAVAILABLE` when asked to do CMAC using 3DES, and log that case as skipped.

This doesn't have any impact on what or how operations get executed. It just allows the self-test to not report a hard failure when an alternative CMAC implementation is compiled in which doesn't support 3DES.

Reply:
This pull request loosens the requirements on a CMAC ALT implementation that has been validated by passing the self-tests. Before, it was required to support 3DES-CMAC. After, it's no longer required. For users of the library, it's a backward incompatible change: if they relied on 3DES-CMAC, and they start running their code on a platform with a CMAC ALT implementation that takes advantage of its newfound freedom to not support 3DES, their code will not work.

I'm ok with this newfound freedom. All I'm asking is to document it with a sentence or so in `cmac.h` and a changelog entry (at least ).

Reply:
This is still not addressed. You've added a note to the CMAC self-test, but that does nothing to inform users of the CMAC module who don't care about the self-tests, and happen to want DES-CMAC. They need to know that if their platform has an ALT implementation of CMAC, it may lack support for DES. Please add a note in the description of `mbedtls_cipher_cmac_starts` and `mbedtls_cipher_cmac` (the two functions that take an argument to choose the underlying cipher) that if `MBEDTLS_CMAC_ALT` is active, `MBEDTLS_CIPHER_DES_EDE3_ECB` may or may not be supported. I think a note in the description of `MBEDTLS_CMAC_C` in `config.h` would be warranted as well.